### PR TITLE
[Automating Model Tracing and Uploading] PR2 Model Listing & Uploading

### PR DESCRIPTION
**Pull request by thanawan-atc**
### Description

Automating model listing updating process with GitHub Actions 
- Add `.github/workflows/update_model_listing.yml` to run `utils/model_uploader/update_pretrained_model_listing.py`  and raise a PR
- Add `utils/model_uploader/update_pretrained_model_listing.py` to update `utils/model_uploader/model_listing/pretrained_model_listing.json` which is used as a source of truth for model listing
- Add `.github/workflows/model_listing_uploader.yml` workflow to upload `pretrained_model_listing.json` to team bucket and then `trigger ml-models-release` jenkins workflow to release the model listing once the PR is closed
- Add tests for `utils/model_uploader/update_pretrained_model_listing.py` in `tests/ml_model_listing` and include it to code coverage

Note:
- The branch used in this automation will be deleted by .github/workflows/delete_model_uploader_branch.yml in PR #209

Related PRs: #209 #211
 
### Check List
- [x] New functionality includes testing.
    - [x] All tests pass
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/source/examples/demo_ml_commons_integration.ipynb [hunk 1]
```diff
@@ -63,21 +64,25 @@
...
-      "Requirement already satisfied: wrapt<2,>=1.10 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from deprecated) (1.15.0)\n",
-      "\u001b[33mDEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621\u001b[0m\u001b[33m\n",
-      "\u001b[0m\n",
-      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.1.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
-      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpython3.9 -m pip install --upgrade pip\u001b[0m\n"
+      "\u001B[33mDEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621\u001B[0m\u001B[33m\n",
```
Comment (dhrubo-os) [inline docs/source/examples/demo_ml_commons_integration.ipynb:75]: I feel like we didn't add any significant changes in this notebook. In that case, can we exclude this from the commit?

File: utils/model_uploader/update_pretrained_model_listing.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    except Exception as e:
+        print(f"Cannot open {filepath} to get model description: {e}")
+        return None
+    if "description" in model_config:
+        return model_config["description"]
+    else:
+        return None
+
+
+def create_new_pretrained_model_listing(
```
Comment (dhrubo-os) [inline utils/model_uploader/update_pretrained_model_listing.py:56]: Is this possible to add unit test for this?

File: docs/source/examples/demo_ml_commons_integration.ipynb (hunk 1 above)
Comment (thanawan-atc) [reply docs/source/examples/demo_ml_commons_integration.ipynb]: Yes, there was a change in PR #208 and this file is outdated in this PR. Sorry for confusion.

File: utils/model_uploader/update_pretrained_model_listing.py (hunk 2 above)
Comment (thanawan-atc) [reply utils/model_uploader/update_pretrained_model_listing.py:55]: @dhrubo-os Tests added. But I don't think it affects code coverage though.

File: utils/__init__.py
Comment (thanawan-atc) [inline utils/__init__.py:1]: @dhrubo-os Can you please verify if this all the license I need?

File: utils/model_uploader/update_pretrained_model_listing.py (hunk 2 above)
Comment (thanawan-atc) [reply utils/model_uploader/update_pretrained_model_listing.py:55]: Just added the file to code coverage.

File: noxfile.py [hunk 3]
```diff
@@ -125,7 +125,7 @@ def test(session, pandas_version: str):
         "-m",
         "pytest",
         "--cov-report=term-missing",
-        "--cov=opensearch_py_ml/",
+        "--cov",
```
Comment (dhrubo-os) [inline noxfile.py:128]: what does it do now? do you see any other concern to change like this?

File: utils/model_uploader/update_pretrained_model_listing.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+import argparse
+import json
+import os
+from typing import Optional
+
+JSON_FILENAME = "pretrained_model_listing.json"
+JSON_DIRNAME = "utils/model_uploader/model_listing"
+PRETRAINED_MODEL_LISTING_JSON_FILEPATH = os.path.join(JSON_DIRNAME, JSON_FILENAME)
+PREFIX_HUGGINGFACE_MODEL_FILEPATH = "ml-models/huggingface/"
```
Comment (dhrubo-os) [inline utils/model_uploader/update_pretrained_model_listing.py:19]: How can we make this extensible?

File: noxfile.py (hunk 3 above)
Comment (thanawan-atc) [reply noxfile.py:128]: By changing it like that, it allows us to specify in the setup.cfg instead what files we want to cover.

Screenshot 2023-08-21 at 1 33 56 PM

File: .github/workflows/update_model_listing.yml [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      uses: actions/checkout@v3
+    - name: Configure AWS Credentials
+      uses: aws-actions/configure-aws-credentials@v2
+      with:
+        aws-region: ${{ secrets.MODEL_UPLOADER_AWS_REGION }}
+        role-to-assume: ${{ secrets.MODEL_UPLOADER_ROLE }}
+        role-session-name: update-model-listing
+    - name: List Models
+      run: |
+        aws s3api list-objects --bucket ${{ secrets.MODEL_BUCKET }} --prefix "ml-models/huggingface/" --query "Contents[].{Key: Key}" --output text | grep "/config.json$" >> config_paths.txt
```
[the remark this answers is not in the document]
Comment (thanawan-atc) [reply .github/workflows/update_model_listing.yml]: Done

File: utils/model_uploader/update_pretrained_model_listing.py (hunk 4 above)
Comment (thanawan-atc) [reply utils/model_uploader/update_pretrained_model_listing.py]: Done

File: .github/workflows/update_model_listing.yml [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    permissions:
+      id-token: write
+      contents: write
+      pull-requests: write
+    environment: opensearch-py-ml-cicd-env
+    env:
+      bucket_model_listing_file_path: ml-models/model_listing/pre_trained_models.json
+      repo_model_listing_path: ./utils/model_uploader/model_listing/pretrained_model_listing.json
+      path_prefixes: "ml-models/huggingface/" 
+      # If you want to the model listing to include more folders you just need to modify path_prefixes
```
Comment (dhrubo-os) [inline .github/workflows/update_model_listing.yml:17]: How about this:


To expand the model listing to encompass additional folders, simply adjust the path_prefixes as indicated below: "ml-models/first_folder/ ml-models/second_folder/ ml-models/third_folder/" (Separate each folder with a space)

File: utils/model_uploader/update_pretrained_model_listing.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+import sys
+from typing import Optional
+
+JSON_FILENAME = "pretrained_model_listing.json"
+JSON_DIRNAME = "utils/model_uploader/model_listing"
+PRETRAINED_MODEL_LISTING_JSON_FILEPATH = os.path.join(JSON_DIRNAME, JSON_FILENAME)
+TORCH_SCRIPT_FORMAT = "TORCH_SCRIPT"
+ONNX_FORMAT = "ONNX"
+TEMP_MODEL_PATH = "temp_model_path"
+TEST_SENTENCES = [
```
Comment (dhrubo-os) [inline utils/model_uploader/update_pretrained_model_listing.py:23]: Where are we using this?

File: utils/model_uploader/update_pretrained_model_listing.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import json
+import os
+import sys
+from typing import Optional
+
+JSON_FILENAME = "pretrained_model_listing.json"
+JSON_DIRNAME = "utils/model_uploader/model_listing"
+PRETRAINED_MODEL_LISTING_JSON_FILEPATH = os.path.join(JSON_DIRNAME, JSON_FILENAME)
+TORCH_SCRIPT_FORMAT = "TORCH_SCRIPT"
+ONNX_FORMAT = "ONNX"
```
Comment (dhrubo-os) [inline utils/model_uploader/update_pretrained_model_listing.py:21]: Where are we using this?

File: utils/model_uploader/update_pretrained_model_listing.py (hunk 7 above)
Comment (thanawan-atc) [reply utils/model_uploader/update_pretrained_model_listing.py]: Just removed it. My bad.

File: utils/model_uploader/update_pretrained_model_listing.py (hunk 8 above)
Comment (thanawan-atc) [reply utils/model_uploader/update_pretrained_model_listing.py]: Same here

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**